pkg/mappings/mappings.go

```diff
@@ -0,0 +1,163 @@
+	Version  string
+	Kind     string
+	Singular string
+	Plural   string
+	Scope    Scope
+}
+
+func (a APIMapping) String() string {
+	return a.Group + "/" + a.Version + " " + a.Plural
+}
+
+// MapperFrom takes a list of APIMappings and uses them to populate a
+// DefaultRESTMapper which can be used locally without connecting to a Discovery
+// client.
+func MapperFrom(mappings []APIMapping) (meta.RESTMapper, []*metav1.APIResourceList) {
+	mapper := meta.NewDefaultRESTMapper(scheme.Scheme.PreferredVersionAllGroups())
+
+	resourceListMapping := map[schema.GroupVersion][]metav1.APIResource{}
+
+	for _, mapping := range mappings {
+		gv := schema.GroupVersion{
+			Group:   mapping.Group,
+			Version: mapping.Version,
+		}
+
+		mapper.AddSpecific(
+			gv.WithKind(mapping.Kind),
+			gv.WithResource(mapping.Plural),
+			gv.WithResource(mapping.Singular),
+			mapping.Scope,
+		)
+
+		resourceListMapping[gv] = append(resourceListMapping[gv], metav1.APIResource{
+			Kind:         mapping.Kind,
+			Name:         mapping.Plural,
+			SingularName: mapping.Singular,
+			Namespaced:   mapping.Scope == "namespace",
+		})
+	}
+
+	resourceLists := []*metav1.APIResourceList{}
+
+	for gv := range resourceListMapping {
+		resourceLists = append(resourceLists, &metav1.APIResourceList{
+			GroupVersion: gv.String(),
+			APIResources: resourceListMapping[gv],
+		})
+	}
+
+	return mapper, resourceLists
+}
+
+//go:embed default-mappings.yaml
+var defaultMappings []byte
+
+// DefaultMapper returns a RESTMapper initialized on default mappings that were
+// embedded into this package. It should include built-in Kubernetes types.
+func DefaultMapper() (meta.RESTMapper, []*metav1.APIResourceList, error) {
+	mappings := []APIMapping{}
+
+	if err := yaml.Unmarshal(defaultMappings, &mappings); err != nil {
+		return nil, nil, err
+	}
+
+	mapper, resList := MapperFrom(mappings)
+
+	return mapper, resList, nil
+}
+
+// GenerateMappings connects to a Kubernetes cluster and discovers the available
+// api-resources, printing out that information as a YAML list of APIMappings.
+// The cluster connected to follows the usual conventions, eg it can be set with
+// the KUBECONFIG environment variable. This function is meant to be used inside
+// of a cobra-style CLI.
+func GenerateMappings(cmd *cobra.Command, _ []string) error {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	clientConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+
+	kubeConfig, err := clientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+
+	client, err := kubernetes.NewForConfig(kubeConfig)
+	if err != nil {
+		return err
+	}
+
+	_, resources, _, err := client.DiscoveryClient.GroupsAndMaybeResources()
+	if err != nil {
+		return err
+	}
+
+	apiMappings := []APIMapping{}
+
+	for gv, resList := range resources {
+		for _, res := range resList.APIResources {
+			if strings.Contains(res.Name, "/") {
+				continue // skip subresources
```

Comment:
@mprahl FYI, `GroupsAndMaybeResources` includes the subresources, which I was ignoring when parsing /api and /apis "by hand"... I didn't see a better way to determine if something was a subresource, but if I included them in the mappings / resources used by `dryrun`, tests would fail and it looked like it was incorrectly checking things in subresources (eg a noncompliant message indicated that a "pod/status did not match")